File: apps/mobile-flutter/test/unit/redirects_test.dart
import 'package:careconnect_mobile/models/session.dart';
import 'package:careconnect_mobile/models/user_role.dart';
import 'package:careconnect_mobile/router/redirects.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  Session session(UserRole role) => Session(
    role: role,
    method: SignInMethod.passkey,
    signedInAt: DateTime(2026, 8, 25),
  );

  group('signed out', () {
    test('everything goes to sign-in except sign-in itself', () {
      expect(resolveRedirect(session: null, location: '/'), '/sign-in');
      expect(
        resolveRedirect(session: null, location: '/patient/today'),
        '/sign-in',
      );
      expect(
        resolveRedirect(session: null, location: '/medications/new'),
        '/sign-in',
      );
      expect(resolveRedirect(session: null, location: '/sign-in'), isNull);
    });
  });

  group('care recipient', () {
    final s = session(UserRole.careRecipient);

    test('lands on Today from the root or sign-in', () {
      expect(resolveRedirect(session: s, location: '/'), '/patient/today');
      expect(resolveRedirect(session: s, location: ''), '/patient/today');
      expect(
        resolveRedirect(session: s, location: '/sign-in'),
        '/patient/today',
      );
    });

    test('may visit patient and shared routes', () {
      expect(
        resolveRedirect(session: s, location: '/patient/medications/med-1'),
        isNull,
      );
      expect(resolveRedirect(session: s, location: '/medications/new'), isNull);
      expect(
        resolveRedirect(session: s, location: '/appointments/a1/edit'),
        isNull,
      );
    });

    test('is bounced away from caregiver screens', () {
      expect(
        resolveRedirect(session: s, location: '/caregiver/dashboard'),
        '/patient/today',
      );
      expect(
        resolveRedirect(session: s, location: '/caregiver/activity'),
        '/patient/today',
      );
    });
  });

  group('caregiver', () {
    final s = session(UserRole.caregiver);

    test('lands on the dashboard and cannot open patient screens', () {
      expect(
        resolveRedirect(session: s, location: '/'),
        '/caregiver/dashboard',
      );
      expect(
        resolveRedirect(session: s, location: '/sign-in'),
        '/caregiver/dashboard',
      );
      expect(
        resolveRedirect(session: s, location: '/patient/today'),
        '/caregiver/dashboard',
      );
      expect(
        resolveRedirect(session: s, location: '/caregiver/manage'),
        isNull,
      );
      expect(resolveRedirect(session: s, location: '/medications/new'), isNull);
    });
  });
}

File: apps/react-mobile/src/navigation/redirects.test.ts
import { resolveRedirect } from './redirects';
import type { Session } from '../models/types';

const patientSession: Session = {
  role: 'careRecipient',
  method: 'passkey',
  signedInAt: new Date(2026, 7, 25),
};

const caregiverSession: Session = {
  role: 'caregiver',
  method: 'email',
  signedInAt: new Date(2026, 7, 25),
  email: 'renee@example.test',
};

describe('resolveRedirect', () => {
  it('sends a signed-out user anywhere to sign-in', () => {
    expect(resolveRedirect({ session: null, location: '/patient/today' })).toBe('/sign-in');
    expect(resolveRedirect({ session: null, location: '/' })).toBe('/sign-in');
  });

  it('allows a signed-out user to stay on sign-in', () => {
    expect(resolveRedirect({ session: null, location: '/sign-in' })).toBeNull();
  });

  it('sends a signed-in user away from sign-in and the root to their home', () => {
    expect(resolveRedirect({ session: patientSession, location: '/sign-in' })).toBe(
      '/patient/today',
    );
    expect(resolveRedirect({ session: patientSession, location: '/' })).toBe('/patient/today');
    expect(resolveRedirect({ session: caregiverSession, location: '' })).toBe(
      '/caregiver/dashboard',
    );
  });

  it('keeps a care recipient off caregiver screens', () => {
    expect(resolveRedirect({ session: patientSession, location: '/caregiver/dashboard' })).toBe(
      '/patient/today',
    );
  });

  it('keeps a caregiver off care-recipient screens', () => {
    expect(resolveRedirect({ session: caregiverSession, location: '/patient/medications' })).toBe(
      '/caregiver/dashboard',
    );
  });

  it('allows a signed-in user to stay on their own role\'s screens', () => {
    expect(resolveRedirect({ session: patientSession, location: '/patient/medications' })).toBeNull();
    expect(resolveRedirect({ session: caregiverSession, location: '/caregiver/manage' })).toBeNull();
  });

  it('allows a signed-in user on a shared, role-agnostic path', () => {
    expect(resolveRedirect({ session: patientSession, location: '/medications/new' })).toBeNull();
  });
});
